server/base/src/main/java/org/apache/accumulo/server/rpc/CustomNonBlockingServer.java

```diff
@@ -144,11 +135,30 @@ public boolean write() {
       boolean result = super.write();
       if (!result) {
         log.trace("CustomFrameBuffer.write returned false when writing data to client: {}",
-            TServerUtils.clientAddress.get());
+            clientAddress);
       }
       return result;
     }
 
+    /*
+     * Helper method used to capture the client address from the socket for each new allocated
+     * FrameBuffer. Because this is a non-blocking server there is a shared thread pool so threads
+     * are re-used for different requests. A new FrameBuffer is created for each request and
+     * different threads can process the same FrameBuffer (For example read() and invoke() can be
+     * called on different threads) so trying to store the client address in a thread local does not
+     * work. The simplest thing to do is to get the client address on FrameBuffer creation as the
+     * same FrameBuffer will be used for the duration of the request.
+     */
+    private String getClientAddress() {
+      String clientAddress = null;
+      if (trans_ instanceof TNonblockingSocket) {
```

Comment:
Is there a way to get the client address for other types?

Reply:
I will look, I just copied what was already there so I didn't check.

Reply:
You do need to cast to TNonblockingSocket as that as the information and parent type doesn't. The abstract type is very sparse and doesn't contain much. Also, this appears to be the only actual implementation anyways so it should be fine.